### PR TITLE
Staked Celo Devchain

### LINKED ISSUE #5: Enable developers to develop against unreleased staked CELO contracts

### LINKED ISSUE #16: Update `@terminal-fi/celo-devchain` dependency once v7 contracts are deployed

### LINKED ISSUE #89: Investigate `execution reverted` error when calling finishingPendingWithdrawals.
When the account contract has multiple pendingWithdrawals, the function fails to execute when attempting to call `lockedGoldWrapper.getPendingWithdrawals(accountContract.address);`

Potential reasons might be:
- concurrent modification exception: When there is a list of withdrawals, the first one is withdrawn, that somehow affects the list, then we move to the next, which is then no longer the second to be withdrawn. Would love to nail this one down. (potentially caused by contract kit)
- Node timeout when reading data from the blockchain
- issue with gas estimation when calling a view function that returns a long list

### PR DESCRIPTION
**Pull request by soloseng**
### Description

Added a script to generate a new devchain tarball containing the deployed stacked-celo contracts.

See `README` on how to use.

### Other changes

In order to add new deployment to devchain tarball, I needed the current celo-devchain package to allow persistant storage of chain data. I made the necessary changes and submitted a [PR](https://github.com/terminal-fi/celo-devchain/issues/2).

In the mean time we are using a github [link](https://github.com/soloseng/celo-devchain/tree/with-db-option) to access the updated celo-devchain package.

### Tested

Added  test script to ensure proper deployment of staked CELO contracts to devchain.

### Related issues

- Fixes #5 
- Fixes #16

### CODE REVIEW

File: package.json
Comment (montera82) [inline package.json:16]: nit: very innovative way to move forward despite delays, and do you think it'll be cool to fork this under the celo-org github organisation?

File: scripts/test/devchain.test.ts
Comment (montera82) [inline scripts/test/devchain.test.ts:25]: any reason why we have these as any, or no typechain for those 🤔

File: README.md
Comment (montera82) [inline README.md:140]: so this is technically checking out a new branch right ..i.e checkout -b name-of-new-branch

File: package.json
Comment (soloseng) [reply package.json:16]: We could do that if we want to maintain our version of the `devchain` npm package. I guess we should evaluate that depending on response time of the original packages maintainers.

File: scripts/test/devchain.test.ts
Comment (soloseng) [reply scripts/test/devchain.test.ts:25]: Since this test is checking the proper deployment of staked CELO contracts on the devchain, for use by anyone in a new project. It assumes the staked CELO contract will not be compiled locally and, therefore, no typings generated before running the test.

File: README.md
Comment (soloseng) [reply README.md:140]: Yes, it's currently checking out a branch that gives us the option to compress the chaindata. (this is temp until branch is merged with `master`.

@m-chrzan thinking more about it, I'm wondering if it might be a bit overkill to clone the `CELO-MONOREPO` only to compress some files vs a script in `STAKED-CELO` repo to handle the compression. Thoughts?

File: README.md
Comment (m-chrzan) [inline README.md:147]: Why are we directing them to save in a `.tmp` directory?

File: README.md
Comment (m-chrzan) [reply README.md:140]: What do you think about having a simple wrapper script here around the monorepo devchain script that takes a `--monorepo` flag (which assumes the monorepo is already built and checked out at the correct branch)?

File: scripts/mineBlocks.ts
Comment (m-chrzan) [inline scripts/mineBlocks.ts:11]: Note there is already a `mineBlocks` function in `test-ts/utils.ts`. Could it be reused?

File: scripts/test/devchain.test.ts
Comment (m-chrzan) [reply scripts/test/devchain.test.ts:25]: But we're running this in this repo, right, so we do have the typings available?

File: scripts/test/devchain.test.ts
Comment (m-chrzan) [inline scripts/test/devchain.test.ts:51]: This looks like an incomplete copy of the previous test?

File: scripts/test/devchain.test.ts
Comment (m-chrzan) [inline scripts/test/devchain.test.ts:61]:
```suggestion
  it("Account should have Manager contract address as manager", async () => {
```

File: scripts/test/devchain.test.ts
Comment (m-chrzan) [inline scripts/test/devchain.test.ts:56]:
```suggestion
  it("Manager should be owned by MultiSig", async () => {
```

File: scripts/test/devchain.test.ts
Comment (m-chrzan) [inline scripts/test/devchain.test.ts:43]:
```suggestion
  it("MultiSig should be owned by 3 addresses", async () => {
```

File: scripts/test/devchain.test.ts
Comment (m-chrzan) [inline scripts/test/devchain.test.ts:27]: `multiSig` here and below

File: scripts/test/devchain.test.ts
Comment (m-chrzan) [inline scripts/test/devchain.test.ts:66]:
```suggestion
  it("StakedCelo should have Manager contract address as manager", async () => {
```

File: README.md
Comment (soloseng) [reply README.md]: Added a script that assumes monorepo is available locally. Details on how to use in the README.

File: scripts/tarchain.ts
Comment (m-chrzan) [inline scripts/tarchain.ts:27]: Doesn't `("" || undefined)` just always evaluate to `undefined`?

File: scripts/tarchain.ts
Comment (m-chrzan) [inline scripts/tarchain.ts:39]: Why is there a `..` before the `monorepo` value? I would assume `monorepo` is meant to point directly at the monorepo already.

File: scripts/tarchain.ts
Comment (m-chrzan) [inline scripts/tarchain.ts:50]: Looks like we're always setting `options` to just a `{ cwd: ... }` object (last line of `runCmd`), is there a code path where `options` is undefined and we end up taking the right side of this `||`? Should all this stuff around `silent` be removed, or should an option for selecting whether we want to run silently or not be added?

File: scripts/test/devchain.test.ts
Comment (m-chrzan) [inline scripts/test/devchain.test.ts:58]: This checks same thing as below test, not that MultiSig is owner.

Comment (m-chrzan) [reply scripts/test/devchain.test.ts:58]: Once this test is correct, similar could be done for Account, StakedCelo, and each Proxy.

File: scripts/tarchain.ts
Comment (soloseng) [reply scripts/tarchain.ts:39]: Used `process.cwd()` instead of `__dirname`. This way `protocolRoot` is based on the `cwd` instead of the location of `tarchain.ts`.

This change should prevent breakage if the `tarchain.ts` get nested further.

File: scripts/tarchain.ts
Comment (soloseng) [reply scripts/tarchain.ts:27]: It never actually evaluates to `undefined`, as `yargs` parses empty `filename` argument as an empty string, but since its type is `string | undefined` I added an undefined check to make it null safe.

File: scripts/tarchain.ts
Comment (soloseng) [reply scripts/tarchain.ts]: Removed the `silent` option, as there is already sufficient error logging without it.

File: scripts/tarchain.ts
Comment (m-chrzan) [reply scripts/tarchain.ts]: I guess I don't understand why you're `join`ing anything to the already fully specified monorepo path. E.g. if `cwd` is `/home/martin/staked-celo` and I provide the monorepo path as an absolute path, e.g. `/home/martin/celo-monorepo`, the current code would lead to `/home/martin/staked-celo/home/martin/celo-monorepo`, right? Which doesn't exist and isn't what we want.

Comment (soloseng) [reply scripts/tarchain.ts]: I joined them since I was using relative path. 

I can change it to using absolute path instead.

Comment (soloseng) [reply scripts/tarchain.ts]: changed to absolute path

Comment (m-chrzan) [reply scripts/tarchain.ts]: I mean, this should now work with either absolute or relative paths. If I do `--monorepo ../<rest of path>` or `--monorepo /<absolute path>`, this should work in either case. Before, it was essentially required you'd provide a relative path, relative to the directory higher up, rather than relative to the current directory. Right?

**Review by m-chrzan — APPROVED** [review]

**Comment by soloseng** [comment]
Made two changes

1. Added a `devchain` network with `port 7545` to the `hardhat.config.ts` and kept `local` with `port 8545` for use with any other local node that we may be running
2. Changed the `tarchain` command to use `--datadir`  & ` --monorepo` flags instead of positional arguments

File: .gitignore
Comment (m-chrzan) [inline .gitignore:11]: This will match the absolute path `/deployments`, no? Which is probably not what we want.

Comment (soloseng) [reply .gitignore:11]: Will we be committing our testnet or mainnet deployment files?

Comment (m-chrzan) [reply .gitignore]: I think that's the convention with hardhat projects, no? You want to keep around the `deployments/network` directory as it keeps states that's necessary for future deployments to run, right?

Comment (soloseng) [reply .gitignore]: Makes sense, specially that we are using upgradable contracts. I made the changes to explicitly ignored the local networks only.

**Pull request merged**
